### PR TITLE
Dont lock numberbox with leading or trailing zeros (T734419)

### PR DESCRIPTION
**Pull request by gvrnmntdept**

### CODE REVIEW

File: js/localization/number.js [hunk 1]
```diff
@@ -299,18 +299,30 @@ var numberLocalization = dependencyInjector({
             cleanedText = text
                 .replace(regExp, "")
                 .replace(decimalSeparator, ".")
-                .replace(/\.$/g, ""),
-            parsed = +cleanedText;
+                .replace(/\.$/g, "");
 
-        cleanedText = cleanedText.replace(/^\./g, "");
+        if(cleanedText === '.' || cleanedText === "") { return null; }
```
Comment (DokaRus) [inline js/localization/number.js]:
```suggestion
        if(cleanedText === "." || cleanedText === "") {
            return null;
        }
```

File: js/localization/number.js [hunk 2]
```diff
@@ -299,18 +299,30 @@ var numberLocalization = dependencyInjector({
...
-        if(cleanedText === "") {
-            return null;
+        for(let index = cleanedText.length - 1; index > separatorIndex; index--) {
+            if(cleanedText.charAt(index) !== '0') {
+                significantDigitsCount += index - separatorIndex;
+                break;
+            }
         }
 
+        if(significantDigitsCount > 15) { return NaN; }
```
Comment (DokaRus) [inline js/localization/number.js]:
```suggestion
        if(significantDigitsCount > 15) {
            return NaN;
        }
```

File: js/localization/number.js [hunk 3]
```diff
@@ -299,18 +299,30 @@ var numberLocalization = dependencyInjector({
...
+        if(cleanedText === '.' || cleanedText === "") { return null; }
 
-        if(cleanedText.length > 15) {
-            return NaN;
+        let separatorIndex = cleanedText.indexOf(".");
+        separatorIndex = separatorIndex < 0 ? cleanedText.length : separatorIndex;
+
+        let significantDigitsCount = 0;
+        for(let index = 0; index < separatorIndex; index++) {
+            if(cleanedText.charAt(index) !== '0') {
```
Comment (DokaRus) [inline js/localization/number.js]:
```suggestion
            if(cleanedText.charAt(index) !== "0") {
```

File: js/localization/number.js [hunk 4]
```diff
@@ -299,18 +299,30 @@ var numberLocalization = dependencyInjector({
...
+                significantDigitsCount += separatorIndex - index;
+                break;
+            }
         }
-        if(cleanedText === "") {
-            return null;
+        for(let index = cleanedText.length - 1; index > separatorIndex; index--) {
+            if(cleanedText.charAt(index) !== '0') {
```
Comment (DokaRus) [inline js/localization/number.js]:
```suggestion
            if(cleanedText.charAt(index) !== "0") {
```

**Review by DokaRus — COMMENTED** [review]
Just a few remarks about the code style

**Review by mrjohn0011 — COMMENTED** [review]
It is OK for me, but i ask @vconst to review this changes. Please wait for his review

**Review by DokaRus — DISMISSED** [review]

**Review by vconst — CHANGES_REQUESTED** [review]
New code contains duplicate and similar lines in two loops. Also, extract new logic into separate method/methods because the parse method becomes big.

File: js/localization/number.js [hunk 5]
```diff
@@ -299,19 +299,28 @@ var numberLocalization = dependencyInjector({
...
-
-        if(cleanedText.length > 15) {
-            return NaN;
-        }
-        if(cleanedText === "") {
+        if(cleanedText === "." || cleanedText === "") {
             return null;
         }
 
+        if(this._calcSignificantDigits(cleanedText) > 15) { return NaN; }
```
Comment (DokaRus) [inline js/localization/number.js]:
```suggestion
        if(this._calcSignificantDigits(cleanedText) > 15) {
            return NaN;
        }
```

File: js/localization/number.js [hunk 6]
```diff
@@ -299,19 +299,28 @@ var numberLocalization = dependencyInjector({
...
+    },
+
+    _calcSignificantDigits: function(text) {
+        const [ integer, fractional ] = text.split(".");
+        const calcDigitsAfterLeadingZeros = digits => {
+            const index = digits.findIndex(x => x !== "0");
+            return index > -1 ? digits.length - index : 0;
+        };
+        let result = 0;
+        if(integer) { result += calcDigitsAfterLeadingZeros(integer.split("")); }
```
Comment (DokaRus) [inline js/localization/number.js]:
```suggestion
        if(integer) {
            result += calcDigitsAfterLeadingZeros(integer.split(""));
        }
```

File: js/localization/number.js [hunk 7]
```diff
@@ -299,19 +299,28 @@ var numberLocalization = dependencyInjector({
...
+        if(fractional) { result += calcDigitsAfterLeadingZeros(fractional.split("").reverse()); }
```
Comment (DokaRus) [inline js/localization/number.js]:
```suggestion
        if(fractional) {
            result += calcDigitsAfterLeadingZeros(fractional.split("").reverse());
        }
```

**Review by DokaRus — DISMISSED** [review]

**Review by DokaRus — APPROVED** [review]

**Review by vconst — APPROVED** [review]
LGTM

**Pull request merged**